tweet_recommendations/other_methods/dbscan_based_method.py

```diff
@@ -0,0 +1,143 @@
+        :param fit_params:
+            minimal_hashtag_occurrence: int. If hashtag occurred less than this number
+            then it's not considered during prediction (simply removed).
+            To include all hashtags put number <= 0.
+            neighbours_count: int. Parameter needed to calculate epsilon param for DBSCAN method.
+            Min param value 2, max param value embedding data count.
+        :return: self.
+        """
+        minimal_hashtag_occurence = fit_params.get("minimal_hashtag_occurence", 10)
+
+        # the value below is so high because otherwise the `epsilon` basing on this would be so small that there would
+        # as many clusters as there are samples in dataset
+        neighbours_count = fit_params.get("neighbours_count", 256)
+
+        if self.verbose:
+            print(f"Data input shape: {x.shape}")
+
+        valid_hashtags = self.drop_tweets_with_hashtags_that_occurred_less_than(x,
+                                                                                minimal_hashtag_occurence)
+        x = self.drop_tweets_which_not_contain_given_hashtags(x, valid_hashtags)
+
+        if self.verbose:
+            print("Setup tweet embedding method")
+
+        self.embedding_method.fit(x)
+
+        if self.verbose:
+            print("Tweet content lemmas embedding started")
+
+        embeddings = self.embedding_method.transform(x["lemmas"])
+
+        if self.verbose:
+            print(f"Data embedding shape, after "
+                  f"droping data by given criteria: {embeddings.shape}")
+
+        self.neighbours = KDTree(embeddings)
+
+        epsilon = np.mean(self.neighbours.query(embeddings, k=neighbours_count, n_jobs=mp.cpu_count())[0])
+
+        if self.verbose:
+            print(f"Found epsilon: {epsilon}")
+            print("Fitting DBSCAN ... ")
+
+        self.clusterizer = DBSCAN(metric='manhattan', eps=epsilon, min_samples=MIN_SAMPLES, n_jobs=mp.cpu_count())
+        x["cluster_label"] = self.clusterizer.fit_predict(embeddings)
+
+        if self.verbose:
+            print("Clustering finished.")
+
+        x["hashtags"] = x["hashtags"].apply(lambda r: [elem["text"] for elem in r])
+
+        grouped_clusters = x.groupby(["cluster_label"])
+
+        if self.verbose:
+            print(f"Samples grouped into {len(np.unique(self.clusterizer.labels_))} clusters")
+            print("Building clusters' data")
+
+        self._centroids_data = []
+        self._corresponding_to_centroids_data_hashtags = []
+
+        for cluster_number, cluster in tqdm.tqdm(grouped_clusters, total=len(np.unique(x["cluster_label"])),
+                                                 disable=not self.verbose):
+            cluster_embeddings = np.vstack(cluster["embedding"].to_numpy())
+            cluster_center = np.mean(cluster_embeddings, axis=0).reshape(1, -1)
+            distances = distance.cdist(cluster_center, cluster_embeddings)[0]  # matrix of shape (1 x N) is returned
```

Comment:
Na pewno klasteryzujemy z użyciem "manhattan" a tutaj liczymy dystans euklidesowy?

Reply:
W publikacji jawnie odnoszą się manhattan do DBScana.
>The DBSCAN algorithm has two parameters: epsilon (The radius of our neighborhoods around a data point p) and minPoints(the minimum number of data points we want in a neighborhood to define a cluster). In our case, we use the Manhattan distance as metric (See eq 2), and for each tweet, we calculate a ...
I dalej ...
> The fact of grouping vectors in this manner is recognized as vector quantization. To perform this, we detect the closest feature vector to the clusters center for each cluster, this vector is considered as the centroid of the cluster.
Także jak dla Ciebie to oznacza stosowanie manhattana, to jasne, mogę zmienić

Reply:
Chyba nie mamy przesłanek, żeby uznać, że tu nagle będzie Euklides, więc raczej zostałbym przy Manhattanie. Brzytwa Ockhama :D